Add support for Shelly Dimmer 2

rnauber opened the pull request:
# What does this implement/fix? 

These patches add support for dimming and power metering with the [Shelly Dimmer 2](https://shelly.cloud/knowledge-base/devices/shelly-dimmer-2/). They are mainly the work of @kostko and use the open-source firmware of @jamesturton for the stm32 microcontroller.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/949

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1753

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:

```yaml


  # the serial port is occupied with the communication to the microcontroller --> disable logging
  logger:
      baud_rate: 0

  uart:
    tx_pin: 1
    rx_pin: 3
    baud_rate: 115200
  
  sensor:

  light:
    - platform: shelly_dimmer
      name: Shelly Dimmer 2 Light
      id: thislight
      power:
        name: Shelly Dimmer 2 Light Power
      voltage:
        name: Shelly Dimmer 2 Light Voltage
      current:
        name: Shelly Dimmer 2 Light Current
      max_brightness: 500

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kostko committed:
shelly_dimmer: Add support for Shelly Dimmer

Commit:
Merge remote-tracking branch 'kostko/kostko/feature/shelly-dimmer' into feature/shelly-dimmer

Commit:
Fix minor API changes of esphome for shelly dimmer.

Commit:
Shelly Dimmer: support multiple firmware versions (from https://github.com/jamesturton/shelly-dimmer-stm32) for the STM microcontroller

Commit:
Shelly Dimmer: add the firmware files from https://github.com/jamesturton/shelly-dimmer-stm32

Commit:
Shelly Dimmer: add a testcase

Commit:
Shelly Dimmer: fix Lint errors (part1)

Commit:
Shelly Dimmer: fix lint part 2

Commit:
Shelly Dimmer: fix lint part3

Commit:
CI: add ".bin" to ignored file types for script/ci-custom.py

Commit:
Shelly Dimmer: fix lint part 4

Commit:
Shelly Dimmer: fix lint part 5

Commit:
Shelly Dimmer: Lint fix part 5

Commit:
Shelly Dimmer: Lint fix part 6

Commit:
Shelly Dimmer: Lint fix part 7

Commit:
Shelly Dimmer: Lint fix part 8

Commit:
Shelly Dimmer: Lint fix part 9

Commit:
Shelly Dimmer: Lint fix part 9

Commit:
Shelly Dimmer: Lint fix part 10

Commit:
Shelly Dimmer: Lint fix part 11

Commit:
Shelly Dimmer: Lint fix part 12

Commit:
Shelly Dimmer: Lint fix part 13

Commit:
Shelly Dimmer: Lint fix part 14

Commit:
Shelly Dimmer: Remove unused options

Commit:
Shelly Dimmer: Lint fix part 15

Commit:
Shelly Dimmer: Lint fix part 16

rnauber commented:
Dear all,
the CI is now almost satisfied, except for two things that I need help (@OttoWinter , @jesserockz , anyone?) with:

1. I inject the firmware into the code as a define (SHD_FIRMWARE_DATA). But this is not recognized by the linter. And the  "NOLINT" annotation is ignored somehow...

```
Error: /home/runner/work/esphome/esphome/esphome/components/shelly_dimmer/shelly_dimmer.cpp:31:40: error: use of undeclared identifier 'SHD_FIRMWARE_DATA' [clang-diagnostic-error]
const uint8_t STM_FIRMWARE[] PROGMEM = SHD_FIRMWARE_DATA;  // NOLINT: SHD_FIRMWARE_DATA is a define
 ``` 

2. The shelly dimmer 2 hardware is ESP8266-only for now. But the `esp32-idf` target is part of the CI and does not compile:

 ``` 
Error: /home/runner/work/esphome/esphome/esphome/components/shelly_dimmer/shelly_dimmer.h:2:10: error: 'HardwareSerial.h' file not found [clang-diagnostic-error]
#include <HardwareSerial.h>
         ^
 ``` 

Should I put all the code under an `#ifdef USE_ESP8266`?


Thanks,
Richard

jesserockz commented:
This is not a review as I have not looked at the code, but:

1. You still need to `define` it somewhere for CI. See `esphome/core/defines.h` which is used only for development and CI.

2. You should not be including the Arduino HardwareSerial lib directly in any component code. All serial stuff you need should be in `esphome/components/uart/uart.h`

Jesse

Commit:
Shelly Dimmer: Add firmware defines to core/defines.h to fix CI.

Commit:
Shelly Dimmer: use eshome's UART implementation

Commit:
Shelly Dimmer: Fix linter (I lost count...)

rnauber commented:
Dear all,
my latest patches to please the linter by using esphome's UART implementation have messed up flashing  the STM. I suspect some subtle (timing?) differences in the implementations. I will look into the problem, meanwhile you can test the component at commit: https://github.com/esphome/esphome/pull/2954/commits/f7de641233ec25342681fc2a4d6816a0fae654a4 .

Best,
Richard

Commit:
Shelly Dimmer: Fix parity in communication with the STM bootloader.

Commit:
Shelly Dimmer: Remove unused code

Commit:
Shelly Dimmer: Fix warning

Commit:
Shelly Dimmer: Fix linter (20ish)

rnauber commented:
Dear all,
I found the problem and it was not subtle at all: I forgot to change the parity to EVEN when flashing the STM microcontroller. As I did not find a mechanism in esphome to reconfigure the UART at runtime (and I am not the only one: https://community.home-assistant.io/t/esphome-change-uart-speed/297737), I fell back to Arduino's HardwareSerial.h implementation just for that.
Now it works for me, plaese test it with your HArdware if you like...

Best,
Richard

Commit:
Shelly Dimmer: fix for the stup@#d linter.

Commit:
Shelly Dimmer: Fix for the linter, even more stupid than the last:)

Commit:
Shelly Dimmer: Expose "warmup_brightness" configuration option.

playya commented:
@oxan which version of the firmware do you use?

playya commented:
I added the uart section to docs PR.
I can confirm that this is working on shelly dimmer 2 and I do not see flickering if I run poweroff animation

DomiStyle commented:
@rnauber You shouldn't use `DEBUG_MSG`, the component won't compile when the logger is set to anything above DEBUG:

```
logger:
  baud_rate: 0
  level: INFO
```

```
src/esphome/components/shelly_dimmer/stm32flash.cpp: In function 'esphome::shelly_dimmer::stm32_err_t esphome::shelly_dimmer::stm32_get_ack_timeout(const stm32_t*, uint32_t)':
src/esphome/components/shelly_dimmer/stm32flash.cpp:165:63: error: expected unqualified-id before ';' token
       DEBUG_MSG(TAG, "Failed to read ACK timeout=%i", timeout);
                                                               ^
```

Replace it with `ESP_LOGD`.

rnauber commented:
> ```
> src/esphome/components/shelly_dimmer/stm32flash.cpp: In function 'esphome::shelly_dimmer::stm32_err_t esphome::shelly_dimmer::stm32_get_ack_timeout(const stm32_t*, uint32_t)':
> src/esphome/components/shelly_dimmer/stm32flash.cpp:165:63: error: expected unqualified-id before ';' token
>        DEBUG_MSG(TAG, "Failed to read ACK timeout=%i", timeout);
>                                                                ^
> ```
> 
> You can replace it with `ESP_LOGD`.

Hm, that is strange as DEBUG_MSG is a macro using ESP_LOGD:
[
`#define DEBUG_MSG(tag, ...) esphome::ESP_LOGD(tag, __VA_ARGS__)` ](https://github.com/rnauber/esphome/blob/9208e35e7735cc0e08cd8954861ae360476ec922/esphome/components/shelly_dimmer/stm32flash.cpp#L30)

If you search and replace DEBUG_MSG with esphome::ESP_LOGD is the problem gone afterwards?

kloknibor commented:
What is the easiest way to test this feature? I have spend some hours trying to setup a development environment to run a locally build version of ESPhome but so far without any luck...

I tried it with the docker build scripts but that just takes the normal esphome lib and I tried gitpod as suggested in the docks which also doesn't work and manually running with python esphome my_configuration.yaml dashboard doesn;t work either since I get an error from the config file...

any help would be appreciated :-)

rnauber commented:
Dear @kloknibor,

> EDIT: I had a very hard time getting it to work on linux... But succeeded using this guide and your PR to get it to work :-) https://shop.sowillo.com/how-to-build-firmware-with-patched-esphome-windows/

That is good to hear! (Another and I think the simplest way to use esphome is as Home Assistant add-on: https://esphome.io/guides/getting_started_hassio.html )

Thank you for testing this PR and please report back your success with the shelly dimmer, maybe even with a video...

Best,
Richard

kloknibor commented:
@rnauber so far I can report back that your PR works perfectly for me! No flickering, good transitions and no bugs found yet. The only issue I still have but that is unrelevant to this PR is to implement dimming on a double click and than stopping the dimming action on another click. (like default with the shelly hardware). I haven;t found a solution to program this in ESPhome so far. But again unrelevant to this PR.

Great work and thanks a lot!

DomiStyle commented:
@rnauber Same here, I have been using your component for 2 weeks now and have no issues to report, everything works as expected except the logging issue.

> If you search and replace DEBUG_MSG with esphome::ESP_LOGD is the problem gone afterwards?
Yes, ESP_LOGD works fine even when using the INFO level.

Commit:
Shelly Dimmer: clean-up logging

Commit:
Shelly Dimmer: fix lint...

Commit:
Shelly Dimmer: formatting

Commit:
Shelly Dimmer: formatting ... again (am I stuck in a loop?)

Commit:
Shelly Dimmer: add myself as codeowner

Commit:
Shelly Dimmer: fix for the linter

rnauber commented:
Hi,
I did some more polishing: I fixed the logging bug that @DomiStyle uncovered and I added myself as code owner (does anyone want to join me, @kostko maybe?). CI/ the linter is satisfied too. So @jesserockz, @OttoWinter, anyone, what else has to be done to get this PR merged? 

Best,
Richard

edge90 commented:
@rnauber 

I'am interested of joining you. I did a lot of the early research regarding protocol, firmware etc way back. I have worked with c++ for over 10 years and has access to a oscilloscope and other debugging tools.

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,478 @@
+}
+
+void ShellyDimmer::setup() {
+  this->buffer_ = new uint8_t[SHELLY_DIMMER_BUFFER_SIZE];  // NOLINT: It will never be freed
```
How about `std::unique_ptr<uint8_t[]>` in the header and `std::make_unique<uint8_t[]>` here?

Edit: or even `std::array<uint8_t, SHELLY_DIMMER_BUFFER_SIZE>` as it will only be 256 (`sizeof(uint8_t)*256`) bytes large. We have 4KB to work with, right?

rnauber replied:
Yes, std::array seems like a good alternative to the c-style array. I am very open to a patch for that:)

edge90 replied:
Absolutely, will do as soon as I have some spare time.

edge90 replied:
https://github.com/rnauber/esphome/pull/37

edge90 commented on esphome/core/defines.h:
```diff
@@ -76,3 +76,9 @@
+// Dummy firmware payload for shelly_dimmer
+#define SHD_FIRMWARE_MAJOR_VERSION 56
+#define SHD_FIRMWARE_MINOR_VERSION 5
+#define SHD_FIRMWARE_DATA \
```
I guess the intension here is to overwrite this maco definition via cli? Overwriting definitons will emit a warning, could be avoided by wrapping this inside `#ifndef .. #endif`

rnauber replied:
No, this file just gives the default defines for unit testing. They should not be overwritten...

edge90 replied:
I'm not sure I'm following. If it's not supposed to be "overriden", then the following line makes no sense? [shelly_dimmer.cpp:34](https://github.com/esphome/esphome/pull/2954/files#diff-f4eb8b9fd43b3d949be24edfe6f9a829491e0f69a29e77ef172404d8712b2f77R34 )

[Example at godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:22,endLineNumber:7,positionColumn:22,positionLineNumber:7,selectionStartColumn:22,selectionStartLineNumber:7,startColumn:22,startLineNumber:7),source:'%0A%23include+%3Ciostream%3E%0A%0A%23define+hello+%22%22%0A%0Aint+main()+%7B%0A++std::cout+%3C%3C+hello%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g112,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O3+-Dhello%3D%22asdasdasd%22',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+11.2+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+11.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

rnauber replied:
There are two cases (mutually exclusive):

1) normal use of esphome: a defines.h is auto generated generated based on the modules you are using in your .yaml.  SHD_FIRMWARE_DATA contains the firmware of the STM microcontroller.

2) unit testing: esphome/core/defines.h is used, SHD_FIRMWARE_DATA contains only dummy values.

So nothing is overwritten and there should be no preprocessor warning (Did you see any warning?).

edge90 replied:
Oh, alright!

rnauber commented:
> @rnauber
> 
> I'am interested of joining you. I did a lot of the early research regarding protocol, firmware etc way back (in the tasmota thread). I have worked with c++ for over 10 years and has access to a oscilloscope and other debugging tools.

That is great news, good to have you on board, @edge90 !

edge90 committed:
Use std::array instead of dynamic allocation

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,478 @@
+namespace esphome {
+namespace shelly_dimmer {
+
+static const char *const TAG = "shelly";
```
```suggestion
static constexpr char TAG[] = "shelly";
```

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,478 @@
+static const char *const TAG = "shelly";
+
+static const uint16_t SHELLY_DIMMER_BUFFER_SIZE = 256;
+static const uint8_t SHELLY_DIMMER_ACK_TIMEOUT = 200;  // ms
```
These could be moved to an anonymous namespace instead of using `static`. Would save some code. `constexpr` could also be used.

rnauber replied:
Could you provide a PR for this too? Thanks!

edge90 replied:
https://github.com/rnauber/esphome/pull/38

rnauber committed:
Merge pull request #37 from edge90/feature/shelly-dimmer-use-std-array

Use std::array instead of dynamic allocation

Commit:
Merge branch 'dev' of https://github.com/esphome/esphome into feature/shelly-dimmer

Commit:
shelly_dimmer: Fixes in the sensor api

rnauber committed:
Update esphome/components/shelly_dimmer/shelly_dimmer.cpp

Co-authored-by: Niclas Larsson <niclas@edgesystems.se>

Commit:
Shelly Dimmer: Fix unused import

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,476 @@
+void ShellyDimmer::send_brightness_(uint16_t brightness) {
+  uint8_t payload[SHELLY_DIMMER_PROTO_CMD_SWITCH_SIZE];
+  // Brightness (%) * 10.
+  payload[0] = brightness & 0xff, payload[1] = brightness >> 8,
```
This must be a bug, right?

`payload[0] = brightness & 0xff, payload[1] = brightness >> 8, this->send_command_(SHELLY_DIMMER_PROTO_CMD_SWITCH, payload, SHELLY_DIMMER_PROTO_CMD_SWITCH_SIZE);`

Is the same as writing:
```
payload[1] = brightness >> 8;
payload[0] = this->send_command_(SHELLY_DIMMER_PROTO_CMD_SWITCH, payload, SHELLY_DIMMER_PROTO_CMD_SWITCH_SIZE);
```

edge90 replied:
See jamesturton's code in [tasmota](https://github.com/arendst/Tasmota/blob/c4bb190e826e0a0b7fb8658bdc9596d7f738857c/tasmota/xdrv_45_shelly_dimmer.ino#L324).

rnauber replied:
Oh yes, you are right that is a nasty bug. I fixed it, thanks!

edge90 committed:
Move constants to anonymous namespace

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,476 @@
+
+/// Computes a crappy checksum as defined by the Shelly Dimmer protocol.
+uint16_t shelly_dimmer_checksum(const uint8_t *buf, int len) {
+  uint16_t sum = 0;
+  for (int i = 0; i < len; i++) {
+    sum += buf[i];
+  }
+  return sum;
```
This is the same as writing:

`#include <numeric>`

and
```suggestion
  return std::accumulate(buf, buf + len, uint16_t{0})
```

Or 
```suggestion
  return std::accumulate<decltype(buf), uint16_t>(buf, buf + len, 0);
```

edge90 reviewed (changes requested):
A lot of stuff is not const, would you like a PR for that as well?

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,476 @@
+  payload[7] = this->warmup_brightness_ >> 8;
+  // Warmup time.
+  payload[8] = this->warmup_time_ & 0xff;
+  payload[9] = this->warmup_time_ >> 8;
```
```suggestion
  uint8_t payload[SHELLY_DIMMER_PROTO_CMD_SETTINGS_SIZE] = {
    // Brightness (%) * 10.
    static_cast<uint8_t>(brightness_int & 0xff),
    static_cast<uint8_t>(brightness_int >> 8),
    // Leading / trailing edge [0x01 = leading, 0x02 = trailing].
    this->leading_edge_ ? uint8_t{0x01} : uint8_t{0x02},
    0x00,
    // Fade rate.
    static_cast<uint8_t>(fade_rate & 0xff),
    static_cast<uint8_t>(fade_rate >> 8),
    // Warmup brightness.
    static_cast<uint8_t>(this->warmup_brightness_ & 0xff),
    static_cast<uint8_t>(this->warmup_brightness_ >> 8),
    // Warmup time.
    static_cast<uint8_t>(this->warmup_time_ & 0xff),
    static_cast<uint8_t>(this->warmup_time_ >> 8),
};
```
If `send_command_` took the payload by const, we could const the array as well (which may catch typos/errors).


An assert could also be used to enforce that all expected values are  there.
```
uint8_t payload[] = {
  ...
};
static_assert(sizeof(payload) == SHELLY_DIMMER_PROTO_CMD_SETTINGS_SIZE);
```

rnauber replied:
Very nice, thank you!

edge90 committed:
Add another space

rnauber committed:
Merge pull request #38 from edge90/feature/refactor-constants

Move constants to anonymous namespace

Commit:
Shelly_dimmer: fix bug in setting the brightness

rnauber committed:
Update esphome/components/shelly_dimmer/shelly_dimmer.cpp

Co-authored-by: Niclas Larsson <niclas@edgesystems.se>

rnauber committed:
Update esphome/components/shelly_dimmer/shelly_dimmer.cpp

Co-authored-by: Niclas Larsson <niclas@edgesystems.se>

rnauber commented:
@edge90 As you are doing all the work, I gave you access to my repository :smile: And please consider adding yourself to the codeowners, If you want. Thank you very much!

edge90 committed:
Add include for numeric

edge90 committed:
const variables

Function signatures as still left to do.

edge90 committed:
static_cast

edge90 committed:
Update codeowner

edge90 committed:
Replace `sizeof(buffer)` with BUFFER_SIZE

edge90 committed:
constexpr STM_FIRMWARE

edge90 committed:
constexpr sizeof call

edge90 committed:
Remove unnecessary variable

edge90 committed:
Include cstring for memcpy

edge90 committed:
Include missing header algorithm

edge90 committed:
Use RAII to close the stm32 uart

edge90 committed:
Move variable to anonymous namespace

edge90 committed:
Rename to size insted of length

We want the size in bytes not the actual number
of elements (even though they're the same in this case).

edge90 committed:
Move around code

* Move static functions to anonymous namespace
* Move stm32_cmd to header

edge90 committed:
Use explicit type as auto isn't allowed until c++14

edge90 committed:
clang-tidy

edge90 committed:
Message for static_assert is required pre c++17

edge90 committed:
Replace std::size with sizeof(p)/sizeof(p[0])

As std::size isn't available until c++17

edge90 committed:
clang-format

edge90 committed:
Run `clang-format -i`

edge90 committed:
Reimplementation of std::size

edge90 committed:
Deduced return type only available with -std=c++14

edge90 committed:
clang-tidy

edge90 committed:
constexpr device table

edge90 commented on esphome/components/shelly_dimmer/dev_table.h:
```diff
@@ -0,0 +1,154 @@
+     0x08040800, 0},
+    {0x9b0, "STM32W-256K", 0x20000200, 0x20004000, 0x08000000, 0x08040000, 4, p_2k, 0x08040800, 0x0804080F, 0x08040000,
+     0x08040800, 0},
+    {0x0}
```
```suggestion
    {0x0, "", 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
```

edge90 committed:
Refactor common code

edge90 committed:
Initialize all

edge90 committed:
Only test static_assert on c++17 or above

edge90 committed:
constexpr instead of defines

edge90 committed:
Clang format namings

edge90 committed:
clang-tidy

edge90 committed:
Missing to initialize some vars

edge90 committed:
Generic code cleanup

edge90 committed:
format

edge90 committed:
More code cleanup

edge90 committed:
Deduced return type doesn't work with c++11

edge90 committed:
Remove code that requires at least c++14

edge90 committed:
even more cleanup

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1050 @@
+  const auto ret = stream->read_array(data, len + 2);
+  if (ret && len == data[0])
+    return STM32_ERR_OK;
+  if (ret == 0 || ret != len + 2) {
```
This doesn't make any sense. I guess the code should check for `if (ret == false)` what do you think?
Reference: https://github.com/stm32duino/stm32flash/blob/main/stm32.c#L327

rnauber replied:
Good catch!
Yes, you are right, read_array returns a bool:
https://esphome.io/api/classesphome_1_1uart_1_1_u_a_r_t_device.html#afe768da23e3af6dc307cb7a011a8f80f
and the check should be `if (ret == false)`.
The old check is a leftover from using the Arduino Serial API previously. Thanks for finding that bug!

edge90 replied:
```suggestion
  if (ret == false) {
```

edge90 committed:
Can't create the buffer const due to hash

edge90 committed:
format once more

edge90 committed:
NOLINT Stm32Dev

edge90 commented:
will update one of my devices in the coming days for testing.

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1035 @@
+  uint8_t buf1[BUFFER_SIZE];
+  populate_buffer_with_address(buf1, address);
+
+  stream->write_array(buf1, 5);
```
```suggestion
  stream->write_array(buf1, BUFFER_SIZE);
```

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1035 @@
+  uint8_t buf[BUFFER_SIZE];
+  populate_buffer_with_address(buf, address);
+
+  stream->write_array(buf, 5);
```
```suggestion
  stream->write_array(buf, BUFFER_SIZE);
```

edge90 committed:
Use constant instead of int literal

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1035 @@
+namespace esphome {
+namespace shelly_dimmer {
+
+namespace {
```
This whole namespace can be moved to a header to make the code less verbose.

ie:
```
namespace {
#include "header.hpp"
}
```

rnauber replied:
Sounds good!

edge90 replied:
These two files would share a lot of the constants which would mean even more files. Not sure it's worth it.

edge90 committed:
Correct bool comparison

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1035 @@
+  const auto ret = stream->read_array(data, len + 2);
+  if (ret && len == data[0])
+    return STM32_ERR_OK;
+  if (ret == false) {
```
```suggestion
  if (!ret) {
```

edge90 committed:
clang-tidy

Commit:
shelly_dimmer: log command's payload

rnauber commented:
I did a quick test with the updated PR and it did work for me (dimming an undimmable 4.5W LED)!

https://user-images.githubusercontent.com/7414650/160071423-91615d6d-80ba-4636-89c5-3037eac2c646.mp4

@edge90 , thanks for the nice refactor! @jesserockz ,do you have advice on what has to be done to get this PR merged? 

Best,
Richard

Bascht74 commented:
@rnauber It would be nice to have an example yaml in the docs… Great work so far! 

do you have a yaml for us right now that we can use to test the component as well?

edge90 committed:
Use RAII as much as possible

edge90 committed:
Yet again, we're using c++11

edge90 committed:
c++11

edge90 commented on esphome/components/shelly_dimmer/light.py:
```diff
@@ -0,0 +1,114 @@
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_VOLTAGE,
+            ),
+            cv.Optional(CONF_CURRENT): sensor.sensor_schema(
```
What do you think about implementing a "auto off" feature based on current? Like the standard shelly dimmer has. This can be done within the config as well but I think it makes sense that a safety feature is always on (and adjusted via option).

rnauber replied:
Having an over-current functionality in the component instead of the yaml makes it more robust/fool-proof. But even better would be an implementation in the STM firmware, as that safety-critical feature would continue to work even if the ESP hangs... But I am fine with both.

edge90 replied:
True, not sure the stm has access to the current data though.

Edit: it has https://github.com/jamesturton/shelly-dimmer-stm32/blob/main/src/main.c#L356 . Agree that the stm firmware is much better place for this

However, the stm32 only checks the current after receiving a poll command from the esp. in a safety context, this is bad.

edge90 replied:
Created https://github.com/jamesturton/shelly-dimmer-stm32/issues/29

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1042 @@
+
+template<typename T> std::unique_ptr<T[], void (*)(T *memory)> malloc_array_raii(size_t size) {
+  // Could be constexpr in c++17
+  static const auto deletor = [](T *memory) { free(memory); };
```
```suggestion
  static const auto DELETOR = [](T *memory) { free(memory); };
```

edge90 committed:
static variable should be uppercase

edge90 committed:
github suggestion misstake

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1042 @@
+   */
+  static constexpr uint32_t MAX_PAGE_SIZE = 512;
+  while (pages) {
+    const auto n = std::max(pages, MAX_PAGE_SIZE);
```
```suggestion
    const auto n = std::min(pages, MAX_PAGE_SIZE);
```

https://github.com/wichers/stm32flash/blob/master/stm32flash.cpp#L894

Bug introduced by me.

edge90 committed:
replace std::max with std::min

edge90 committed:
clang-tidy

rnauber commented:
Oh, since my last test something changed that prevented flashing the STM32: 

```
[22:38:06][W][shelly:072]: Unsupported STM32 firmware version, flashing
[22:38:06][W][shelly:121]: Starting STM32 firmware upgrade
[22:38:06][D][shelly:443]: Reset STM32, boot0=1
[22:38:06][D][shelly:460]: Reset STM32 done
[22:38:14][I][mqtt:214]: MQTT Connected!
[22:38:14][I][shelly:062]: Initializing Shelly Dimmer...
[22:38:14][D][shelly:443]: Reset STM32, boot0=0
[22:38:14][D][shelly:460]: Reset STM32 done
[22:38:14][D][shelly:236]: Sending command: 0x11 (0 bytes) payload 0x
[22:38:14][D][shelly:248]: Command sent, waiting for reply
[22:38:14][D][shelly:358]: Got frame: 0x11
[22:38:14][I][shelly:068]: STM32 current firmware version: 51.5, desired version: 51.6
[22:38:14][W][shelly:072]: Unsupported STM32 firmware version, flashing
[22:38:14][W][shelly:121]: Starting STM32 firmware upgrade
[22:38:14][D][shelly:443]: Reset STM32, boot0=1
[22:38:14][D][shelly:460]: Reset STM32 done
[22:38:22][I][mqtt:214]: MQTT Connected!
[22:38:22][I][shelly:062]: Initializing Shelly Dimmer...
[22:38:22][D][shelly:443]: Reset STM32, boot0=0
[22:38:22][D][shelly:460]: Reset STM32 done
[22:38:22][D][shelly:236]: Sending command: 0x11 (0 bytes) payload 0x
[22:38:22][D][shelly:248]: Command sent, waiting for reply
[22:38:22][D][shelly:358]: Got frame: 0x11
[22:38:22][I][shelly:068]: STM32 current firmware version: 51.5, desired version: 51.6
```

rnauber commented:
> @rnauber It would be nice to have an example yaml in the docs… Great work so far!
> 
There is a snippet here:
https://github.com/esphome/esphome-docs/pull/1753
> do you have a yaml for us right now that we can use to test the component as well?
Yes, have a look here:
https://github.com/esphome/feature-requests/issues/949#issuecomment-1002682045

Best,
Richard

edge90 committed:
Forgot to release the unique_ptr

edge90 commented:
@rnauber can you try again. Stupid mistake of mine.

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1053 @@
+
+  // TODO: Would be much better if the unique_ptr was returned from this function
+  // Release ownership of unique_ptr
+  stm_raii.release();
```
```suggestion
  static_cast<void*>(stm_raii.release());
```

edge90 committed:
Suppress warning

rnauber commented:
> @rnauber can you try again? Stupid mistake of mine.

In 7ae844e404e1d381cddd4e54e8c0c850f02ca566 flashing works. Thanks for fixing this, @edge90 .

```
[22:33:37][I][app:133]: Rebooting safely...
[22:33:53][I][mqtt:214]: MQTT Connected!
[22:33:53][I][shelly:062]: Initializing Shelly Dimmer...
[22:33:53][D][shelly:443]: Reset STM32, boot0=0
[22:33:53][D][shelly:460]: Reset STM32 done
[22:33:53][D][shelly:236]: Sending command: 0x11 (0 bytes) payload 0x
[22:33:53][D][shelly:248]: Command sent, waiting for reply
[22:33:53][D][shelly:358]: Got frame: 0x11
[22:33:53][I][shelly:068]: STM32 current firmware version: 51.6, desired version: 51.5
[22:33:53][W][shelly:072]: Unsupported STM32 firmware version, flashing
[22:33:53][W][shelly:121]: Starting STM32 firmware upgrade
[22:33:53][D][shelly:443]: Reset STM32, boot0=1
[22:33:53][D][shelly:460]: Reset STM32 done
[22:33:55][I][shelly:171]: STM32 firmware upgrade successful
[22:33:55][D][shelly:443]: Reset STM32, boot0=0
[22:33:55][D][shelly:460]: Reset STM32 done
[22:33:55][D][shelly:236]: Sending command: 0x11 (0 bytes) payload 0x
[22:33:55][D][shelly:248]: Command sent, waiting for reply
[22:33:55][D][shelly:358]: Got frame: 0x11
[22:33:55][I][shelly:068]: STM32 current firmware version: 51.5, desired version: 51.5
[22:33:55][D][shelly:208]: Brightness update: 0 (raw: 0.000000)
[22:33:55][D][shelly:236]: Sending command: 0x20 (10 bytes) payload 0x0000020000002c011400
[22:33:55][D][shelly:248]: Command sent, waiting for reply
[22:33:55][D][shelly:358]: Got frame: 0x20
[22:33:55][D][shelly:236]: Sending command: 0x01 (2 bytes) payload 0x0000
[22:33:55][D][shelly:248]: Command sent, waiting for reply
[22:33:55][D][shelly:358]: Got frame: 0x01
[22:33:55][D][shelly:236]: Sending command: 0x10 (0 bytes) payload 0x
[22:33:55][D][shelly:248]: Command sent, waiting for reply
[22:33:55][D][shelly:358]: Got frame: 0x10
```

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1053 @@
+
+  // TODO: Would be much better if the unique_ptr was returned from this function
+  // Release ownership of unique_ptr
+  static_cast<void*>(stm_raii.release());
```
```suggestion
  static_cast<void *>(stm_raii.release());
```

edge90 committed:
format

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1053 @@
+
+  // TODO: Would be much better if the unique_ptr was returned from this function
+  // Release ownership of unique_ptr
+  static_cast<void *>(stm_raii.release());
```
```suggestion
  static_cast<void *>(stm_raii.release());  // NOLINT
```

edge90 committed:
suppress clang-diagnostic-unused-value

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+  this->reset_dfu_boot_();
+
+  // Could be constexpr in c++17
+  const auto close = [](stm32_t *stm32) { stm32_close(stm32); };
+
+  // Cleanup with RAII
+  std::unique_ptr<stm32_t, decltype(close)> stm32{stm32_init(this, STREAM_SERIAL, 1), close};
```
```suggestion
  static const auto CLOSE = [](stm32_t *stm32) { stm32_close(stm32); };

  // Cleanup with RAII
  std::unique_ptr<stm32_t, decltype(CLOSE)> stm32{stm32_init(this, STREAM_SERIAL, 1), CLOSE};
```

edge90 committed:
Make lambda static

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+    break;
+  }
+
+  // Poll the dimmer roughly every 10s.
+  this->set_interval("poll", 10000, [this]() { this->send_command_(SHELLY_DIMMER_PROTO_CMD_POLL, nullptr, 0); });
+
+  this->send_settings_();
+  // Do an immediate poll to refresh current state.
+  this->send_command_(SHELLY_DIMMER_PROTO_CMD_POLL, nullptr, 0);
```
When you use `set_interval` the function is scheduled to run in the next loop for the first time, so this will run almost straight after setup is done.

Also might be better to convert this to a `PollingComponent` maybe which will take care of this?

rnauber replied:
Done!

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+    return 0;
+  }
+
+  uint16_t brightness_int = static_cast<uint16_t>(brightness * (this->max_brightness_ - this->min_brightness_));
+  brightness_int = brightness_int + this->min_brightness_;
+  return std::min(brightness_int, SHELLY_DIMMER_MAX_BRIGHTNESS);
```
`SHELLY_DIMMER_MAX_BRIGHTNESS` should be removed and the config should be validated to only allow max 1000 for `CONF_MAX_BRIGHTNESS`

```suggestion
  return remap<uint16_t, float>(brightness, 0.0f, 1.0f, this->min_brightness_, this->max_brightness_);
```

edge90 replied:
Reference: https://esphome.io/api/helpers_8h_source.html#l00145

edge90 replied:
```suggestion
  return remap<uint16_t, float>(brightness, 0.0f, 1.0f, this->min_brightness_, this->max_brightness_);
```

jesserockz replied:
Haha woops, got those template types the wrong way around.

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+      // payload[1] is unused.
+      const uint16_t brightness = payload[3] << 8 | payload[2];
+
+      const uint32_t power_raw = payload[7] << 24 | payload[6] << 16 | payload[5] << 8 | payload[4];
+
+      const uint32_t voltage_raw = payload[11] << 24 | payload[10] << 16 | payload[9] << 8 | payload[8];
+
+      const uint32_t current_raw = payload[15] << 24 | payload[14] << 16 | payload[13] << 8 | payload[12];
```
Just to reuse the helpers we have =)

```suggestion
      const uint32_t power_raw = encode_uint32(payload[7], payload[6], payload[5], payload[4]);

      const uint32_t voltage_raw = encode_uint32(payload[11], payload[10], payload[9], payload[8]);

      const uint32_t current_raw = encode_uint32(payload[15], payload[14], payload[13], payload[12]);
```

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+
+      const uint8_t hw_version = payload[0];
+      // payload[1] is unused.
+      const uint16_t brightness = payload[3] << 8 | payload[2];
```
```suggestion
      const uint16_t brightness = encode_uint16(payload[3], payload[2]);
```

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+#include "esphome/core/defines.h"
```
For some of the below comments
```suggestion
#include "esphome/core/defines.h"
#include "esphome/core/helpers.h"
```

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -0,0 +1,490 @@
+
+namespace {
+
+constexpr char TAG[] = "shelly";
```
For consistency with codebase
```suggestion
static const char *const TAG = "shelly_dimmer";
```

edge90 replied:
Makes more sense to have constexpr as it is a compile time constant. By changing the type from `char[]` to `char*` you'll loose the size known at compile time.

rnauber replied:
I am fine with both! :)

jesserockz commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1053 @@
+*/
+
+#include <cstdint>
+#include <Arduino.h>
```
You have some `#ifndef USE_ESP_IDF` code in the above files, but then have this with `#include <Arduino>` 
Are we able to drop the Arduino requirement from this? Alternatively drop the flashing support for ESP_IDF if it cannot be modified. (See main review comment too)

rnauber replied:
Oh, that is just a leftover. I will drop it!

jesserockz reviewed (changes requested):
A few things below, but my major request for change is that we make the entire firmware flashing optional and only compile it in when a user specifies the version in yaml **and** a path to the firmware file. 
This would mean that ESPHome does not need to keep the binary files inside the repo and releases and a user can source the files from the official source and flash them if they want to.

I don't like that everyday use currently means that the firmware binary is always going to be compiled in and would prefer that people just add the firmware config if and when they want to change the firmware.

WDYT?

edge90 committed:
Use encode_uintXX helpers

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

edge90 committed:
use remap for calculating brightness

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

edge90 commented:
@rnauber 

How about?

firmware:
  type: github|file|whatever
  url: ... | version: "51.5"

jesserockz commented:
Downloading from GitHub can be done in python sure, I have no issues with that.

I am happy for you to figure out the options etc, I just want it to be so that by default it wont compile in the binaries and flashing code. The user has to specify some sort of config to have that code compiled in.

Maybe a schema like this:
```py
cv.Optional("firmware"): cv.maybe_simple_value(
  {
    cv.Optional("file"): validate_local_file_or_url, # This needs to be made
    cv.Required("version"): cv.one_of(*FIRMWARE_MAPPING), # maybe?
  },
  validate_file, # This would convert a simple version key to generate the full url
  key="version",
)
```

This would allow for:
```yaml
firmware: "51.5"
```

```yaml
firmware: 
  url: https:///
  version: 51.5
```

Or just leaving it out completely which would not compile in any of the flashing code, but would still check the version to make sure it is compatible with known versions.

jesserockz commented:
> Are there any helpers for URL handling/downloading that I should use?

@rnauber Take a look at the way google fonts are downloaded and cached

https://github.com/esphome/esphome/blob/dev/esphome/components/font/__init__.py#L124-L158

Commit:
Shelly Dimmer: download firmware from an url and validate the hash

Commit:
Shelly Dimmer: make flashing of the firmware optional (and non-default)

Commit:
Shelly Dimmer: remove firmware binaries from repo

Commit:
Shelly Dimmer: Linted

Commit:
Shelly Dimmer: Linted again

Commit:
Shelly Dimmer: Linted again

Commit:
Shelly Dimmer: Linted again

Commit:
Shelly Dimmer: drop ARDUINO.H

Commit:
Shelly Dimmer: use PollingComponent

rnauber commented:
Hi everyone,
I think all suggestions of  @jesserockz and @edge90 are implemented now. The biggest change is that flashing the firmware is optional now (and not including the firmware saves a few bytes on the ESP). I tested both cases, and it works. 
I settled for this scheme:

```
            cv.Optional(CONF_FIRMWARE, default="51.6"): cv.maybe_simple_value(
                {
                    cv.Optional(CONF_URL): cv.url,
                    cv.Optional(CONF_SHA256): validate_sha256,
                    cv.Required(CONF_VERSION): validate_version,
                    cv.Optional(CONF_UPDATE, default=False): cv.boolean,
                },
                validate_firmware,  # converts a simple version key to generate the full url
                key=CONF_VERSION,
            ),
```
So what do you think?

Best,
Richard

jesserockz commented on esphome/components/shelly_dimmer/light.py:
```diff
@@ -43,12 +46,108 @@
+    return [HexInt(x) for x in firmware_data]
+
+
+def validate_firmware(value):
+    if CONF_URL not in value:
+        try:
+            value[CONF_URL], value[CONF_SHA256] = KNOWN_FIRMWARE[value[CONF_VERSION]]
+        except KeyError as e:
+            raise cv.Invalid(
+                f"Firmware {value[CONF_VERSION]} is unknown, please specify an '{CONF_URL}' ..."
+            ) from e
+    return value
```
This should copy the object and return the copy
```suggestion
def validate_firmware(value):
    config = value.copy()
    if CONF_URL not in config:
        try:
            config[CONF_URL], config[CONF_SHA256] = KNOWN_FIRMWARE[config[CONF_VERSION]]
        except KeyError as e:
            raise cv.Invalid(
                f"Firmware {config[CONF_VERSION]} is unknown, please specify an '{CONF_URL}' ..."
            ) from e
    return config
```

jesserockz commented on esphome/components/shelly_dimmer/light.py:
```diff
@@ -75,22 +174,26 @@
-    firmware_data = Path(firmware_path).read_bytes()
-    hexdata = [HexInt(x) for x in firmware_data]
-    cg.add_define("SHD_FIRMWARE_DATA", hexdata)
+    fw_hex = get_firmware(config[CONF_FIRMWARE])
```
`get_firmware` should be run as a validator on the schema (it can be put after `validate_firwmware` or even part of that function itself), if it throws `cv.Invalid` here it is a crash and does not have the nice config errors that Invalid can produce.

rnauber replied:
Ok. And fw_hex should be passed through the config?

jesserockz replied:
Well the fw_hex can be loaded from the local file within `to_code` as you can assume the validators have ensured the file is downloaded by that point.

rnauber replied:
Done!

rnauber replied:
Done

jesserockz commented on esphome/components/shelly_dimmer/light.py:
```diff
@@ -75,22 +174,26 @@
+    if fw_hex is not None:
+        cg.add_define("SHD_FIRMWARE_DATA", fw_hex)
     cg.add_define("SHD_FIRMWARE_MAJOR_VERSION", fw_major)
     cg.add_define("SHD_FIRMWARE_MINOR_VERSION", fw_minor)
```
Please prefix these with `USE_` to match other defines we use in ESPHome
```suggestion
        cg.add_define("USE_SHD_FIRMWARE_DATA", fw_hex)
    cg.add_define("USE_SHD_FIRMWARE_MAJOR_VERSION", fw_major)
    cg.add_define("USE_SHD_FIRMWARE_MINOR_VERSION", fw_minor)
```

jesserockz commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -19,7 +19,6 @@
 */
 
 #include <cstdint>
-#include <Arduino.h>
```
Please also wrap this cpp and the h file in `#ifdef SHD_FIRMWARE_DATA`

rnauber replied:
Done

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -86,20 +93,24 @@ void ShellyDimmer::setup() {
       // Firmware upgrade completed, do the checks again.
       continue;
+#else
+      ESP_LOGW(TAG, "Firmware version mismatch, put 'update: True' in the yaml to flash an update.");
```
This message might also be good to put in `dump_config` as most people would miss the setup logs if they are connecting wirelessly.
```suggestion
      ESP_LOGW(TAG, "Firmware version mismatch, put 'update: true' in the yaml to flash an update.");
```

rnauber replied:
Ok, I added a dump_config method.

jesserockz reviewed (changes requested):
Mostly looking good, just a few things below.

rnauber committed:
Update esphome/components/shelly_dimmer/light.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

rnauber committed:
Update esphome/components/shelly_dimmer/shelly_dimmer.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

Commit:
Shelly Dimmer: add dump_config

Commit:
Shelly Dimmer: add USE_ prefix to firmware defines

Commit:
Shelly Dimmer: add "#ifdef USE_SHD_FIRMWARE_DATA" to stm32flash.cpp

Commit:
Shelly Dimmer: download firmware as part of the validation

Commit:
Shelly Dimmer: Lint...

jesserockz commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -111,6 +112,30 @@ void ShellyDimmer::setup() {
+  LOG_PIN("  NRST Pin: ", this->pin_nrst_);
+  LOG_PIN("  BOOT0 Pin: ", this->pin_boot0_);
+
+  ESP_LOGCONFIG(TAG, "  Leading Edge: %b", this->leading_edge_);
```
This is in `esphome/core/log.h`
https://github.com/esphome/esphome/blob/dev/esphome/core/log.h#L165-L167
```suggestion
  ESP_LOGCONFIG(TAG, "  Leading Edge: %s", YESNO(this->leading_edge_));
```

jesserockz commented on script/ci-custom.py:
```diff
@@ -80,7 +80,7 @@ def find_all(a_str, sub):
 )
 cpp_include = ("*.h", "*.c", "*.cpp", "*.tcc")
-ignore_types = (".ico", ".woff", ".woff2", "")
+ignore_types = (".ico", ".woff", ".woff2", ".bin", "")
```
Not needed anymore
```suggestion
ignore_types = (".ico", ".woff", ".woff2", "")
```

jesserockz commented on esphome/components/shelly_dimmer/stm32flash.h:
```diff
@@ -0,0 +1,124 @@
+*/
+
+#pragma once
+
```
```suggestion

#include "esphome/core/defines.h"
#ifdef USE_SHD_FIRMWARE_DATA

```

jesserockz commented on esphome/components/shelly_dimmer/stm32flash.h:
```diff
@@ -0,0 +1,124 @@
+uint32_t stm32_sw_crc(uint32_t crc, uint8_t *buf, unsigned int len);
+
+}  // namespace shelly_dimmer
+}  // namespace esphome
```
```suggestion
}  // namespace esphome

#endif  // USE_SHD_FIRMWARE_DATA
```

rnauber committed:
Update esphome/components/shelly_dimmer/shelly_dimmer.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

rnauber committed:
Update script/ci-custom.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

rnauber committed:
Update esphome/components/shelly_dimmer/stm32flash.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

rnauber committed:
Update esphome/components/shelly_dimmer/stm32flash.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

Commit:
Shelly Dimmer: add #ifdef to dev_table.h

edge90 commented on esphome/components/shelly_dimmer/dev_table.h:
```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "esphome/core/defines.h"
+#ifdef USE_SHD_FIRMWARE_DATA
```
This file is only used by stm32flash.cpp. How about not having the ifdef guard here?

jesserockz replied:
This should be added to any file we do not want to compile the contents of with our define flags, which includes this file.

rnauber replied:
Yes, and it did not compile otherwise...

edge90 replied:
A header that compiles? Strange indeed.

edge90 committed:
Rename tag to shelly_dimmer

jesserockz reviewed (approved):
I think this looks good to me now.

edge90 reviewed (approved):
Looks good to me as well. The only thing I would like to change is that `stm32_init` should return the unqiue_ptr instead of releasing it. That could however be done at a later stage.

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1056 @@
+  uint8_t buf[257];
+
+  auto *stm = (stm32_t *) calloc(sizeof(stm32_t), 1);      // NOLINT
+  stm->cmd = (stm32_cmd_t *) malloc(sizeof(stm32_cmd_t));  // NOLINT
```
These two lines may return a nullptr.

Bascht74 commented:
As the next beta version is closing by: Is there a chance to get this thing into the next beta (scheduled for tomorrow, if I understand the pattern right)?

edge90 commented on esphome/components/shelly_dimmer/stm32flash.cpp:
```diff
@@ -0,0 +1,1056 @@
+
+  if (stm->cmd->get == STM32_CMD_ERR || stm->cmd->gvr == STM32_CMD_ERR || stm->cmd->gid == STM32_CMD_ERR) {
+    ESP_LOGD(TAG, "Error: bootloader did not returned correct information from GET command");
+    return nullptr;
+  }
+
+  /* get the device ID */
+  if (stm32_guess_len_cmd(stm, stm->cmd->gid, buf, 1) != STM32_ERR_OK) {
+    return nullptr;
+  }
+  const auto returned = buf[0] + 1;
+  if (returned < 2) {
+    ESP_LOGD(TAG, "Only %d bytes sent in the PID, unknown/unsupported device", returned);
+    return nullptr;
+  }
+  stm->pid = (buf[1] << 8) | buf[2];
+  if (returned > 2) {
+    ESP_LOGD(TAG, "This bootloader returns %d extra bytes in PID:", returned);
+    for (auto i = 2; i <= returned; i++)
+      ESP_LOGD(TAG, " %02x", buf[i]);
+  }
+  if (stm32_get_ack(stm) != STM32_ERR_OK) {
+    return nullptr;
+  }
+
+  stm->dev = DEVICES;
+  while (stm->dev->id != 0x00 && stm->dev->id != stm->pid)
+    ++stm->dev;
+
+  if (!stm->dev->id) {
+    ESP_LOGD(TAG, "Unknown/unsupported device (Device ID: 0x%03x)", stm->pid);
+    return nullptr;
+  }
+
+  // TODO: Would be much better if the unique_ptr was returned from this function
+  // Release ownership of unique_ptr
+  static_cast<void *>(stm_raii.release());  // NOLINT
+
+  return stm;
+}
```
The malloc and calloc can leak and/or segfault with the current code.

```suggestion
stm32_t *stm32_init(uart::UARTDevice *stream, const uint8_t flags, const char init) {
  uint8_t buf[257];

  // Could be constexpr in c++17
  static const auto CLOSE = [](stm32_t *stm32) { stm32_close(stm32); };

  // Cleanup with RAII
  std::unique_ptr<stm32_t, decltype(CLOSE)> stm{static_cast<stm32_t *>(calloc(sizeof(stm32_t), 1)),  // NOLINT
                                                CLOSE};

  if (!stm) {
    return nullptr;
  }

  stm->cmd = static_cast<stm32_cmd_t *>(malloc(sizeof(stm32_cmd_t)));  // NOLINT
  if (!stm->cmd) {
    return nullptr;
  }

  /* get the version and read protection status  */
  if (stm32_send_command(stm.get(), STM32_CMD_GVR) != STM32_ERR_OK) {
    return nullptr;  // NOLINT
  }

  memset(stm->cmd, STM32_CMD_ERR, sizeof(stm32_cmd_t));
  stm->stream = stream;
  stm->flags = flags;

  if ((stm->flags & STREAM_OPT_CMD_INIT) && init) {
    if (stm32_send_init_seq(stm.get()) != STM32_ERR_OK)
      return nullptr;  // NOLINT
  }

  /* From AN, only UART bootloader returns 3 bytes */
  {
    const auto len = (stm->flags & STREAM_OPT_GVR_ETX) ? 3 : 1;
    if (!stream->read_array(buf, len))
      return nullptr;  // NOLINT
    stm->version = buf[0];
    stm->option1 = (stm->flags & STREAM_OPT_GVR_ETX) ? buf[1] : 0;
    stm->option2 = (stm->flags & STREAM_OPT_GVR_ETX) ? buf[2] : 0;
    if (stm32_get_ack(stm.get()) != STM32_ERR_OK) {
      return nullptr;
    }
  }

  {
    const auto len = ([&]() {
      /* get the bootloader information */
      if (stm->cmd_get_reply) {
        for (auto i = 0; stm->cmd_get_reply[i].length; ++i) {
          if (stm->version == stm->cmd_get_reply[i].version) {
            return stm->cmd_get_reply[i].length;
          }
        }
      }

      return STM32_CMD_GET_LENGTH;
    })();

    if (stm32_guess_len_cmd(stm.get(), STM32_CMD_GET, buf, len) != STM32_ERR_OK)
      return nullptr;
  }

  const auto stop = buf[0] + 1;
  stm->bl_version = buf[1];
  int new_cmds = 0;
  for (auto i = 1; i < stop; ++i) {
    const auto val = buf[i + 1];
    switch (val) {
      case STM32_CMD_GET:
        stm->cmd->get = val;
        break;
      case STM32_CMD_GVR:
        stm->cmd->gvr = val;
        break;
      case STM32_CMD_GID:
        stm->cmd->gid = val;
        break;
      case STM32_CMD_RM:
        stm->cmd->rm = val;
        break;
      case STM32_CMD_GO:
        stm->cmd->go = val;
        break;
      case STM32_CMD_WM:
      case STM32_CMD_WM_NS:
        stm->cmd->wm = newer(stm->cmd->wm, val);
        break;
      case STM32_CMD_ER:
      case STM32_CMD_EE:
      case STM32_CMD_EE_NS:
        stm->cmd->er = newer(stm->cmd->er, val);
        break;
      case STM32_CMD_WP:
      case STM32_CMD_WP_NS:
        stm->cmd->wp = newer(stm->cmd->wp, val);
        break;
      case STM32_CMD_UW:
      case STM32_CMD_UW_NS:
        stm->cmd->uw = newer(stm->cmd->uw, val);
        break;
      case STM32_CMD_RP:
      case STM32_CMD_RP_NS:
        stm->cmd->rp = newer(stm->cmd->rp, val);
        break;
      case STM32_CMD_UR:
      case STM32_CMD_UR_NS:
        stm->cmd->ur = newer(stm->cmd->ur, val);
        break;
      case STM32_CMD_CRC:
        stm->cmd->crc = newer(stm->cmd->crc, val);
        break;
      default:
        if (new_cmds++ == 0) {
          ESP_LOGD(TAG, "GET returns unknown commands (0x%2x", val);
        } else {
          ESP_LOGD(TAG, ", 0x%2x", val);
        }
    }
  }
  if (new_cmds)
    ESP_LOGD(TAG, ")");
  if (stm32_get_ack(stm.get()) != STM32_ERR_OK) {
    return nullptr;
  }

  if (stm->cmd->get == STM32_CMD_ERR || stm->cmd->gvr == STM32_CMD_ERR || stm->cmd->gid == STM32_CMD_ERR) {
    ESP_LOGD(TAG, "Error: bootloader did not returned correct information from GET command");
    return nullptr;
  }

  /* get the device ID */
  if (stm32_guess_len_cmd(stm.get(), stm->cmd->gid, buf, 1) != STM32_ERR_OK) {
    return nullptr;
  }
  const auto returned = buf[0] + 1;
  if (returned < 2) {
    ESP_LOGD(TAG, "Only %d bytes sent in the PID, unknown/unsupported device", returned);
    return nullptr;
  }
  stm->pid = (buf[1] << 8) | buf[2];
  if (returned > 2) {
    ESP_LOGD(TAG, "This bootloader returns %d extra bytes in PID:", returned);
    for (auto i = 2; i <= returned; i++)
      ESP_LOGD(TAG, " %02x", buf[i]);
  }
  if (stm32_get_ack(stm.get()) != STM32_ERR_OK) {
    return nullptr;
  }

  stm->dev = DEVICES;
  while (stm->dev->id != 0x00 && stm->dev->id != stm->pid)
    ++stm->dev;

  if (!stm->dev->id) {
    ESP_LOGD(TAG, "Unknown/unsupported device (Device ID: 0x%03x)", stm->pid);
    return nullptr;
  }

  // TODO: Would be much better if the unique_ptr was returned from this function
  // Release ownership of unique_ptr
  return stm.release();  // NOLINT
}
```

rnauber replied:
Oh no, that change broke flashing:

```
07:32:34][D][shelly_dimmer:479]: Reset STM32, boot0=0
[07:32:34][D][shelly_dimmer:496]: Reset STM32 done
[07:32:34][D][shelly_dimmer:272]: Sending command: 0x11 (0 bytes) payload 0x
[07:32:34][D][shelly_dimmer:284]: Command sent, waiting for reply
[07:32:34][D][shelly_dimmer:394]: Got frame: 0x11
[07:32:34][I][shelly_dimmer:075]: STM32 current firmware version: 51.6, desired version: 51.5
[07:32:34][W][shelly_dimmer:081]: Unsupported STM32 firmware version, flashing
[07:32:34][W][shelly_dimmer:158]: Starting STM32 firmware upgrade
[07:32:34][D][shelly_dimmer:479]: Reset STM32, boot0=1
[07:32:34][D][shelly_dimmer:496]: Reset STM32 done
[07:32:42][I][mqtt:214]: MQTT Connected!
[07:32:42][I][shelly_dimmer:069]: Initializing Shelly Dimmer...
[07:32:42][D][shelly_dimmer:479]: Reset STM32, boot0=0
[07:32:42][D][shelly_dimmer:496]: Reset STM32 done
[07:32:42][D][shelly_dimmer:272]: Sending command: 0x11 (0 bytes) payload 0x
[07:32:42][D][shelly_dimmer:284]: Command sent, waiting for reply
[07:32:42][D][shelly_dimmer:394]: Got frame: 0x11
[07:32:42][I][shelly_dimmer:075]: STM32 current firmware version: 51.6, desired version: 51.5
[07:32:42][W][shelly_dimmer:081]: Unsupported STM32 firmware version, flashing
[07:32:42][W][shelly_dimmer:158]: Starting STM32 firmware upgrade
[07:32:42][D][shelly_dimmer:479]: Reset STM32, boot0=1
[07:32:42][D][shelly_dimmer:496]: Reset STM32 done
[07:32:50][I][mqtt:214]: MQTT Connected!
[07:32:50][I][shelly_dimmer:069]: Initializing Shelly Dimmer...
[07:32:50][D][shelly_dimmer:479]: Reset STM32, boot0=0
[07:32:50][D][shelly_dimmer:496]: Reset STM32 done
[07:32:50][D][shelly_dimmer:272]: Sending command: 0x11 (0 bytes) payload 0x
[07:32:50][D][shelly_dimmer:284]: Command sent, waiting for reply
[07:32:50][D][shelly_dimmer:394]: Got frame: 0x11
[07:32:50][I][shelly_dimmer:075]: STM32 current firmware version: 51.6, desired version: 51.5
[07:32:50][W][shelly_dimmer:081]: Unsupported STM32 firmware version, flashing
[07:32:50][W][shelly_dimmer:158]: Starting STM32 firmware upgrade
[07:32:50][D][shelly_dimmer:479]: Reset STM32, boot0=1
[07:32:50][D][shelly_dimmer:496]: Reset STM32 done
```
and the CI complains:
```

### File esphome/components/shelly_dimmer/stm32flash.cpp

/home/runner/work/esphome/esphome/esphome/components/shelly_dimmer/stm32flash.cpp:183:3: error: Called C++ object pointer is null [clang-analyzer-core.CallAndMessage,-warnings-as-errors]
  stream->write_array(buf, BUFFER_SIZE);
  ^
```


As it would be nice to have this PR included in the upcoming beta version, should I revert this patch and we fix it later?

edge90 replied:
@rnauber I'm sorry, most have copied to much or little. Could you please try again?

rnauber committed:
Update esphome/components/shelly_dimmer/stm32flash.cpp

Co-authored-by: Niclas Larsson <niclas@edgesystems.se>

edge90 committed:
Init comes first

rnauber commented:
Thanks @edge90, it works flawlessly, now!  
I would declare this PR ready, so thanks to all who contributed!

Best,
Richard
